### ISSUE TITLE
Gtk2 and Glib2 reactor on Python 2.5 and earlier cause problems with interrupted syscalls

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4286]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#4286|
|Type|defect|
|Created|2010-02-21 18:40:26Z|

[#10460](https://github.com/twisted/twisted/issues/10460) fixed most of Twisted to play nicely with respect to signal handlers and interrupted syscalls, but failed to fix the problem for Gtk2 and Glib2 reactor.

This could be fixed by having these reactors explicitly select `twisted.internet._sigchld` as the way to handle `SIGCHLD`.

<details><summary>Searchable metadata</summary>

```
trac-id__4286 4286
type__defect defect
reporter__exarkun exarkun
priority__low low
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1266777626000000 1266777626000000
changetime__1348858392000000 1348858392000000
version__None None
owner__itamar itamar

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

I don't understand why the gtk2 and glib2 reactors have these problems.  What is it that the gtk mainloop does differently?  Does importing 'gtk' or 'gobject' install signal handlers?

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst commented|
|-|-|

This has become relevant to [#5512](https://github.com/twisted/twisted/issues/5512), maybe. In addition to Glyph's question - what is the purpose of _GlibSignalMixin at all? It seems like it still sets a SIGCHLD handler *before* gtk would run its event loop, so if gtk.main would override it what's the point?

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst commented|
|-|-|

I re-encountered this in [#5986](https://github.com/twisted/twisted/issues/5986). Looking at current glib code, it looks like we just stomp on any use by gtk/glib of the set_wakeup_fd. Maybe that's OK, inspection of the code suggests they only use it to detect Ctrl-C, and don't actually look at contents of the byte. So in so far as we manage shutdown, it's not a problem.

http://git.gnome.org/browse/pygtk/tree/gtk/gtk.override?id=PYGTK_2_24_0#n1146

I therefore propose closing this as wontfix. Any disagreement?

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst set owner to @itamarst|
|-|-|

The relevant test passes just fine on the gtk2 and glib2 reactors, JP has no comment, Glyph is silent. I will close this as part of [[#5986](https://github.com/twisted/twisted/issues/5986)](https://github.com/twisted/twisted/issues/5986), since I'll already be deleting skips in this test made irrelevant by #5986.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph commented|
|-|-|

Well, we don't support python 2.5 anyway, right?  So a fix would have to be a patch version, and it doesn't sound like anyone cares.  So go ahead.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_7"></a>|@itamarst set status to `closed`|
|-|-|

(In [35835]) Merge signals-5986-3.

Author: itamar
Review: exarkun
Fixes: [#5986](https://github.com/twisted/twisted/issues/5986), [#4286](https://github.com/twisted/twisted/issues/4286)

Port twisted.internet._signals to Python 3, dropping unnecessary pre-2.6 code along the way.
